sync/upstream.py

```diff
@@ -462,6 +462,65 @@ def pr_head(self):
         ref = self.git_wpt.refs[pr_ref]
         return ref.commit.hexsha
 
+    @property
+    def pr_commits(self):
+        pr_head = self.pr_head
+        if not pr_head:
+            raise ValueError("Can't get PR commits as the ref head could not be found for %s" %
+                             self.process_name)
+        else:
+            pr_head = sync_commit.WptCommit(self.git_wpt, pr_head)
+
+        merge_base = []
+
+        # Check if the PR Head is reachable from origin/master
+        origin_master_sha = self.git_wpt.refs['origin/master'].commit.hexsha
+        pr_head_reachable = self.git_wpt.is_ancestor(pr_head.sha1, origin_master_sha)
+
+        # If not reachable, then it either hasn't landed yet, it was a Squash + Merge,
+        # or a Rebase and merge.
+        if not pr_head_reachable:
+            # TODO: how to effectively detect Rebase and Merge?
```

Comment:
I'm still a bit unsure of how to detect this scenario, other than checking this merge base against the one returned by GH, but even then I am unsure which merge base it would give (gut feeling says the new merge base)

Reply:
We don't need to detect how the actual merge was done as such, we only need to care about the final state of the tree, so I'm not sure exactly what the concern is here.

Reply:
I guess I was concerned about if this is used to determine which changes to apply to central, we may end up applying the same changes twice, but yes for the purposes of finding the commits in the PR it has no relevance.

Reply:
If we have a situation where we have PR commits like `[from gecko, from gecko, fixup]` and we land with a rebase we can obviously just ignore the first two when we come to do the landing. If they're landed with a squash, we can probably compare `pr_commits` with `wpt_commits`, notice that there's one new commit and just apply that. If there's something like `[from gecko but modified]` we probably just have to stick with the current approach of applying everything and hoping there aren't conflicts. But that's far fewer cases.